### PR TITLE
Add hooks for InputConnection lock and unlocking

### PR DESCRIPTION
**Pull request by mklim**
With this, plugins can know whether or not their input connection should
be cached. In very rare cases this can be used by plugins to know how to
more properly handle their own input connections, in cases where they're
overriding typical input behavior themselves.

This adds a hook needed for flutter/flutter#19718

### CODE REVIEW

**Comment by matthew-carroll** [comment]
@mklim can you describe the issue that this is resolving, e.g., why would a plugin need or want to cache an input connection? And it looks like changes are to platform views-specific logic, is there a reason that plugins would only be operating on platform views?

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java [hunk 1]
```diff
@@ -366,6 +366,30 @@ public View getPlatformViewById(Integer id) {
         return controller.getView();
     }
 
+    /**
+    * Callback fired when the platform's input connection is locked, or should be used. See also
+    * {@link TextInputPlugin#lockPlatformViewInputConnection}.
+    */
+    public void onInputConnectionLocked(int viewId) {
```
Comment (amirh) [inline shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: instead of adding these methods and triggering this from TextInputPlugin we could directly trigger this from resizePlatformView.

File: shell/platform/android/io/flutter/plugin/platform/PlatformView.java [hunk 2]
```diff
@@ -24,4 +25,26 @@
...
+    /**
+    * Callback fired when the platform's input connection is locked, or should be used. See also
+    * {@link TextInputPlugin#lockPlatformViewInputConnection}.
+    *
+    * <p>This hook only exists for rare cases where the plugin relies on the state of the input
+    * connection. This probably doesn't need to be implemented.
+    */
+    // Default interface methods are supported on all min SDK versions of Android.
+    @SuppressLint("NewApi")
+    default void onInputConnectionLocked() {};
```
Comment (amirh) [inline shell/platform/android/io/flutter/plugin/platform/PlatformView.java]: [not sure about this, what do you think?]

Alternatively we could call these lockInputConnection/unlockInputConnection, these hooks allows us to move the locking logic out of TextInputPlugin and into the keyboard using platform view, in which case I think the alternative names are better. OTOH I'm not sure if we don't want to provide the locking behavior by default as this should work for any platform view that doesn't do the crazy threading tricks webview is doing.

**Comment by mklim** [comment]
> can you describe the issue that this is resolving, e.g., why would a plugin need or want to cache an input connection?

I think the related issue has the full context here: flutter/flutter#19718. But to try and briefly sum it up: a plugin would specifically want to cache an input connection if and when they're deliberately doing strange things with input connections themselves and then need to handle the times when it _should_ be cached correctly. The specific use case that's driving this is that embedded webviews on Android pre N exhibit some behavior that means that by default text fields inside the embedded webview just "don't work." The webview plugin needs to manually manage input connections in the plugin itself to work around this and correctly set up the right input connection for the webview.

> And it looks like changes are to platform views-specific logic, is there a reason that plugins would only be operating on platform views?

Sorry, yeah. This hook isn't really for plugins in general, but for plugins with platform views that need to manage input connections manually. It's extremely specific and unlikely to be used by plugins in general.

**Comment by amirh** [comment]
> @mklim can you describe the issue that this is resolving, e.g., why would a plugin need or want to cache an input connection? And it looks like changes are to platform views-specific logic, is there a reason that plugins would only be operating on platform views?

Android WebView is running the input connection on a dedicated thread due to performance reasons. As doing this is only supported starting Android N, prior to Android N webview is pulling a trick that makes the IMM invoke onCreateInputConnection on an unattached view on their dedicated thread, when this trick is performed following attempts to create an input connection do not hit FlutterView, which essentially bypasses our current locking mechanism implemented in TextInputPlugin. The hooks added in this PR allows the WebView plugin to address the issue by implementing it's own caching mechanism.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 1 above)
Comment (mklim) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: I like having to avoid the extra plumbing, but I don't like having the actual call back into the plugin be separated from the real bit in `TextInputPlugin`. Moving the logic there feels brittle to me, like one could get called without the other and then things would end up out of sync. WDYT? I could be being overly cautious.

File: shell/platform/android/io/flutter/plugin/platform/PlatformView.java (hunk 2 above)
Comment (mklim) [reply shell/platform/android/io/flutter/plugin/platform/PlatformView.java]: I don't think I understand what you're suggesting, can you go into more detail? Are you saying we should implement logic here by default or is this comment talking about moving this call into `PlatformViewsController#resize`?

Assuming you're talking about shifting logic here, into the `default` methods, I'd rather avoid that if possible. I think doing that would have similar readability issues with base classes, where there's secret behavior up the hierarchy chain that's kind of hard to understand and easily broken accidentally.

Comment (amirh) [reply shell/platform/android/io/flutter/plugin/platform/PlatformView.java]: I'm talking about not doing any locking in the engine code, but leave it up to the plugin.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 1 above)
Comment (amirh) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: I see your point, I do not expect there to ever be another call site for lock/unlock so my personal preference would be to avoid the plumbing at the cost of not having a mechanism enforcing the coupling of these calls.

**Comment by matthew-carroll** [comment]
Sounds good on the explanation of the need. You might consider actually placing an explicit scenario explanation like this within the source code docs. While source docs are generally framed generically because they can be used for many use-cases, it might be difficult for future readers to grok the nature of this behavior, and locating this PR in GitHub history would be a friction that might cause readers to simply ignore the implications of the logic.

File: shell/platform/android/io/flutter/plugin/platform/PlatformView.java [hunk 3]
```diff
@@ -24,4 +25,26 @@
...
     void dispose();
+
+    /**
+    * Callback fired when the platform's input connection is locked, or should be used. See also
+    * {@link TextInputPlugin#lockPlatformViewInputConnection}.
+    *
+    * <p>This hook only exists for rare cases where the plugin relies on the state of the input
+    * connection. This probably doesn't need to be implemented.
+    */
+    // Default interface methods are supported on all min SDK versions of Android.
```
Comment (matthew-carroll) [inline shell/platform/android/io/flutter/plugin/platform/PlatformView.java]: Do we plan to add this comment not to every default method? If not, then we probably shouldn't add it to any.

File: shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java (hunk 1 above)
Comment (matthew-carroll) [reply shell/platform/android/io/flutter/plugin/platform/PlatformViewsController.java]: IMO, bypassing TextInputPlugin for a text input related behavior sounds like an architectural issue waiting to happen. I think TextInputPlugin should be involved in all behaviors that impact Android text input related to a Flutter experience. I don't think we can say with confidence where any such requirements will go in the future....

**Comment by amirh** [comment]
@matthew-carroll (for some reason I cannot reply nor quote your last comment).

There are Android text input behaviors that are purely there to provide text input support for platform views, lock/unlock is one of them, I do not think we should just say that any text input related behavior must live in the TextInputPlugin e.g should the focus tricks we do in the presentation live in the text input plugin? technically I'm sure we could add a bunch of plumbing to move the relevant code there.

**Comment by matthew-carroll** [comment]
@amirh I originally agreed with that sentiment, but the more time I've spent in the embedding the more I've reversed position on that. I think it's important that we avoid treating anything that looks similar to an extension as a sacred cow. Even platform views. I understand that from a certain perspective they seem core to Flutter because any given Flutter dev can use them without declaring a new library, but actually I think that they're functionally closer to plugins than they are to core behavior. For example, text input, touch input, and accessibility are all behaviors that I would consider core. But platform views, 1P and 3P plugins, and background execution containers are all things that I think can be treated as extensions of Flutter. Flutter really makes no sense for anyone without text input, touch input, and accessibility, but Flutter still makes perfect sense for many apps without platform views and background execution.

In other words, I don't think platform views should be able to do anything or access anything that other developers are prohibited from using, nor should platform views hide any special relationships. I think this lends itself to Flutter's philosophy of a layered and modular architecture. I also think it is likely to put us in a better position for unknowable future requirements.

I welcome thoughts from @mklim and @xster on this question of philosophy as well.

**Comment by amirh** [comment]
> @amirh I originally agreed with that sentiment, but the more time I've spent in the embedding the more I've reversed position on that. I think it's important that we avoid treating anything that looks similar to an extension as a sacred cow. Even platform views.

Are we treating platform views as a sacred cow?

> I understand that from a certain perspective they seem core to Flutter because any given Flutter dev can use them without declaring a new library, but actually I think that they're functionally closer to plugins than they are to core behavior. For example, text input, touch input, and accessibility are all behaviors that I would consider core. But platform views, 1P and 3P plugins, and background execution containers are all things that I think can be treated as extensions of Flutter. Flutter really makes no sense for anyone without text input, touch input, and accessibility, but Flutter still makes perfect sense for many apps without platform views and background execution.

Flutter makes perfect sense for many apps without text input and other things you may consider "core" as well.

> 
> In other words, I don't think platform views should be able to do anything or access anything that other developers are prohibited from using,

What do you mean by "prohibited"? unavailable to code that is external to the embedder / engine? we can technically take any component, and expose enough hooks to implement it out of the core engine / embedder, probably even the rasterizer. At some point it's impractical, platform views on iOS for example changes the compositor to composite to multiple surface, would you expose hooks to do that in code that lives out of the engine? 

Also this discussion is completely unrelated to the topic discussed in this PR.

> nor should platform views hide any special relationships.

What special relationship is being hidden here?
The PlatformView interface is the interface that a plugin uses to communicate with the platform views mechanism, having a component call a method on an interface it defines isn't a privileged access or hiding a relationship, any component can call methods on an interface it defined.

---

I see the original topic as a tradeoff call between simplicity and safety. As I said before my personal opinion is that in this case having the PlatformViewsController call the TextInputPlugin#lock just so it calls the PlatformViewsController#onLock is an overkill given the amount of added boilerplate and that this lock mechanism is used only for and by the PlatformViewsController. My preference on that consideration is not very strong, not strong enough to spend much time debating over it. Still, I feel like the current code is more complex and harder to follow than it needs to be.

**Comment by matthew-carroll** [comment]
> > @amirh I originally agreed with that sentiment, but the more time I've spent in the embedding the more I've reversed position on that. I think it's important that we avoid treating anything that looks similar to an extension as a sacred cow. Even platform views.
> 
> Are we treating platform views as a sacred cow?

I believe we are. I believe we are blurring the boundary between platform views and text input/accessibility, and I think that we're doing it because we're on the inside of the embedding and therefore we can. But that doesn't necessarily mean we should.
> 
> > I understand that from a certain perspective they seem core to Flutter because any given Flutter dev can use them without declaring a new library, but actually I think that they're functionally closer to plugins than they are to core behavior. For example, text input, touch input, and accessibility are all behaviors that I would consider core. But platform views, 1P and 3P plugins, and background execution containers are all things that I think can be treated as extensions of Flutter. Flutter really makes no sense for anyone without text input, touch input, and accessibility, but Flutter still makes perfect sense for many apps without platform views and background execution.
> 
> Flutter makes perfect sense for many apps without text input and other things you may consider "core" as well.

I don't actually think it's accurate to compare apps that don't receive text input to apps that don't use platform views. You can correct me if I'm wrong, but I would guess the difference in volume is an order of magnitude or more. I do think there is a reasonable argument to be made that text input, touch input, and accessibility are core in a manner that platform views are not. Even FlutterView, FlutterFragment, and FlutterActivity, despite being used all the time, have a clear API boundary between them and the FlutterEngine related classes. So what I'm describing isn't specific to platform views. Very few things need to be, or should be tightly coupled to the central behaviors of rendering, I/O, and accessibility.
> 
> > In other words, I don't think platform views should be able to do anything or access anything that other developers are prohibited from using,
> 
> What do you mean by "prohibited"? unavailable to code that is external to the embedder / engine? we can technically take any component, and expose enough hooks to implement it out of the core engine / embedder, probably even the rasterizer. At some point it's impractical, platform views on iOS for example changes the compositor to composite to multiple surface, would you expose hooks to do that in code that lives out of the engine?

Isn't your suggestion in this PR that the platform views system take control of the input connection under certain circumstances, and to do so without the text input plugin being involved? If so, that is a situation where a responsibility that is owned by the text input plugin is temporarily extracted, handled by something else, and then eventually given back. I don't think any external developers could do something like that, and I think it could make it very difficult in the future for other developers to understand where certain responsibilities belong. But maybe I misunderstood the recommendation?
> 
> Also this discussion is unrelated to the topic discussed in this PR.

The relationship is that if platform views does something with text input that is unknown and/or uncontrolled by the object responsible for text input (TextInputPlugin), then I think we are giving special privileges to the platform views system that it probably shouldn't have, and doing so in a way that external developers could not replicate, which I'm suggesting is something we should try very hard to avoid.
> 
> > nor should platform views hide any special relationships.
> 
> What special relationship is being hidden here?
> The PlatformView interface is the interface that a plugin uses to communicate with the platform views mechanism, having a component call a method on an interface it defines isn't a privileged access or hiding a relationship, any component can call methods on an interface it defined.
> 
> I see the original topic as a tradeoff call between simplicity and safety. As I said before my personal opinion is that in this case having the PlatformViewsController call the TextInputPlugin#lock just so it calls the PlatformViewsController#onLock is an overkill given the amount of added boilerplate and that this lock mechanism is used only for and by the PlatformViewsController. My preference on that consideration is not very strong, not strong enough to spend much time debating over it. Still, I feel like the current code is more complex and harder to follow than it needs to be.

I think that perhaps the difference in perspective here is that what you're labeling boilerplate, I would label an encapsulation boundary. I don't think it's just extra stuff that we have to deal with. It's a cut line between one subsystem and another. I think it's those specific types of APIs that keep components modular and ensure that we don't lose the ability to talk about and reason about these subsystems independently.

It's true that what I'm describing goes beyond what is in this PR. I've also shared a doc that is related to this topic as well. There is a question of how we approach an entire category of issue, but it appears that what is being described in this PR is a member of that category (as far as I can tell), so I wanted to raise the issue. I'm happy to continue in another medium, but I would like for us to try to reach some kind of alignment so that we know what to expect in the future and we can avoid spending time on similar debates.

Also, I really would appreciate input from @mklim and @xster on the philosophy I'm describing because I think having 3+ perspectives is more helpful in triangulating than just 2. I think it would be good if all 4 of us were on the same page moving forward as we're all likely to review each others' work in this area for some time.

**Comment by amirh** [comment]
> @matthew-carroll @amirh I do think the broader architectural discussions on the purpose of `TextInputPlugin` and the relative popularity of `PlatformView`s are a little orthogonal to this patch, maybe we should catch up offline or in another thread? I can see how it's related to the change but I think the particular question regarding this patch is more local.
> 
> It comes down to a style question outside of broader architecture. On one hand there's the mild risk of duplication of state in multiple places to avoid plumbing the state from point A to B. On the other there's a bunch of plumbing code for the sake of keeping the state ultimately in one callsite. I've looked into this more to try and form a concrete opinion on it beyond just my own preferences and [not duplicating state](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-duplicating-state) is mentioned in the style guide as something to stick with. So with that in mind I do think that it's better to keep them connected, even at the cost of the boilerplate.
> 
> @amirh does that rationale make sense to you? Happy to talk about this more.

My preference here is not super strong, so I'm fine with whatever you decide.

Offline discussion - this isn't a question of duplicate state vs not duplicating state, the "locked" state is duplicated between the WebView plugin and the TextInputPlugin in both cases.

**Comment by amirh** [comment]
> Isn't your suggestion in this PR that the platform views system take control of the input connection under certain circumstances, and to do so without the text input plugin being involved? If so, that is a situation where a responsibility that is owned by the text input plugin is temporarily extracted, handled by something else, and then eventually given back.

No, the platform views system is just signaling the (webview) plugin that the next time it's asked for an input connection it should return the currently active one.

**Comment by mklim** [comment]
Discussed offline. The locked bit in TextInputPlugin only exists in order to track when `resize()` was last called by platform views, so in this case it does make sense to just call back to plugin from there. Updated the PR with that suggestion. The code smells here point to broader architectural questions as mentioned above that the four of us should catch up on at a later point.
